src/wallet/scriptpubkeyman.cpp

```diff
@@ -357,9 +357,8 @@ void LegacyScriptPubKeyMan::MarkUnusedAddresses(const CScript& script)
         if (it != mapKeyMetadata.end()){
             CKeyMetadata meta = it->second;
             if (!meta.hd_seed_id.IsNull() && meta.hd_seed_id != m_hd_chain.seed_id) {
-                bool internal = (meta.key_origin.path[1] & ~BIP32_HARDENED_KEY_LIMIT) != 0;
+	    	bool internal = meta.key_origin.path[1] == (1 | 0x80000000);
```

Comment:
> Personal preference: the former logic is more defensive.

Glad you brought this up.  I also thought to make mainnet path more restrictive but I used line from `walletdb.cpp` instead assuming it's acceptable to be this way.  Perhaps this change will make mainnet path more restrictive (taking into consideration legacy version path):

`bool internal = ((meta.key_origin.path[1] == 0x80000003) || (meta.key_origin.path[1] == 0x80000000)) ? 0 : 1; `

(i.e. testnet for everything else except when path[1] is 3 or 0).

Reply:
The way it was (and the way you suggest in the comment) is more protective of mainnet than testnet: I think that is better because when a wallet with an unknown path shows up, assuming that it is mainnet is riskier than assuming it's testnet. I also think - but have to test it - that the fallback to having a wallet with an incompatible path can simply be restored and rescanned into a descriptor wallet rather than a "legacy wallet". Thinking of it, this is actually important to document somewhere because it's bound to happen at some point - I've added it to the 1.21 tracker issue.

The readability of this code as-is is already far from optimal, but I cannot complain because I didn't contribute to it on Bitcoin Core. It needs documentation and maybe a tiny refactor which we can do on top. Since those are best done without changing any logic (implement the same result just in a better way), how about you change this to your proposed code from your comment and then I'll do a little refactor/doc pull on the top? That way, we have a clean path.